### PR TITLE
Dev eve output redis async

### PR DESCRIPTION
**Pull request by fooinha**
Initial support for reading redis replies asynchronously.
The libevent adapter for the hiredis async API is used.

Submission for review. @jasonish 

Command pipelining for the async: true mode is not implemented, and it only will be added after discussion and review. 

Thanx

### CODE REVIEW

**Comment by inliniac** [comment]
Can you describe the high level design a bit more?

cc: @regit

File: src/util-logopenfile.c [hunk 1]
```diff
@@ -592,26 +863,35 @@ static int  LogFileWriteRedis(LogFileCtx *file_ctx, const char *string, size_t s
...
+                default:
+                    SCLogError(SC_ERR_INVALID_VALUE,
+                            "Redis default triggered with %d", reply->type);
+                    SCConfLogReopenRedis(file_ctx);
+                    break;
+            }
+
+            freeReplyObject(reply);
+        }  else {
+            if (unlikely(reply == NULL)) {
```
Comment (regit) [inline src/util-logopenfile.c]: we are in else of if reply test so this test is not necessary.

File: configure.ac [hunk 2]
```diff
@@ -1869,6 +1869,52 @@
...
+                echo
+                echo "   ERROR!  libevent library not found, go get it"
+                echo
+                echo "   Ubuntu: apt-get install libevent-dev"
+                echo "   Fedora: dnf install libevent-devel"
+                echo "   RHEL/CentOS: yum install libevent-devel"
+                echo
+                exit 1
+            fi
+            if test "$LIBEVENT" = "yes"; then
```
Comment (regit) [inline configure.ac]: Could be merged with test on l. 1894

File: configure.ac [hunk 3]
```diff
@@ -1869,6 +1869,52 @@
         fi
     fi
 
+#libevent if hiredis is enabled
+    AC_ARG_ENABLE(libevent,
+            AS_HELP_STRING([--enable-libevent],[Enable libevent for async calls for redis ]),
```
Comment (regit) [inline configure.ac]: White space before ]

Comment (regit) [inline configure.ac]: I don't like the idea of having a new flag in configure. Could we have it at least in auto mode so if hiredis if selected then we try to get libevent ? and fallback to no libevent if not available.

File: src/util-logopenfile.c [hunk 4]
```diff
@@ -34,9 +34,34 @@
...
 #include "util-logopenfile-tile.h"
 
+#ifdef HAVE_LIBHIREDIS
+#include <hiredis/hiredis.h>
+
 const char * redis_push_cmd = "LPUSH";
 const char * redis_publish_cmd = "PUBLISH";
 
+
+#if HAVE_LIBEVENT == 1
```
Comment (regit) [inline src/util-logopenfile.c]: Please use `#ifdef HAVE_LIBEVENT` has done in other define tests.

File: src/util-logopenfile.c [hunk 5]
```diff
@@ -190,11 +215,11 @@ SCLogOpenFileFp(const char *path, const char *append_setting)
  *  \retval FILE* on success
  *  \retval NULL on error
  */
-static PcieFile *SCLogOpenPcieFp(LogFileCtx *log_ctx, const char *path,
```
Comment (regit) [inline src/util-logopenfile.c]: Don't mix whitespace fixes with a functional commit.

File: src/util-logopenfile.c [hunk 6]
```diff
@@ -190,11 +215,11 @@ SCLogOpenFileFp(const char *path, const char *append_setting)
  *  \retval FILE* on success
  *  \retval NULL on error
  */
-static PcieFile *SCLogOpenPcieFp(LogFileCtx *log_ctx, const char *path, 
+static PcieFile *SCLogOpenPcieFp(LogFileCtx *log_ctx, const char *path,
                                  const char *append_setting)
 {
 #ifndef __tile__
-    SCLogError(SC_ERR_INVALID_YAML_CONF_ENTRY,
```
Comment (regit) [inline src/util-logopenfile.c]: Idem.

File: src/util-logopenfile.c [hunk 7]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+/** \brief SCLogRedisContextAlloc() - Allocates and initalizes redis context
+ *  \param async indicates that async mode will be used
+ *  \retval SCLogRedisContext * pointer if succesful, EXIT_FAILURE program if not
+ */
+static SCLogRedisContext * SCLogRedisContextAlloc(int async)
+{
+
+    SCLogRedisContext* ctx = (SCLogRedisContext*) SCMalloc(sizeof(SCLogRedisContext));
+
+	if (unlikely(ctx == NULL)) {
```
Comment (regit) [inline src/util-logopenfile.c]: Indentation is wrong

File: src/util-logopenfile.c [hunk 8]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+        SCLogError(SC_ERR_MEM_ALLOC, "Unable to allocate redis context!");
+        exit(EXIT_FAILURE);
+	}
+
+	ctx->sync    = NULL;
```
Comment (regit) [inline src/util-logopenfile.c]: code style please.

File: src/util-logopenfile.c [hunk 9]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+    /* Disconnection or lost reply may have happened */
+    if (reply == NULL) {
+        if (ctx->ev_base != NULL) {
+            event_base_loopbreak(ctx->ev_base);
+        }
+        SCConfLogReopenRedis(file_ctx);
+        return;
+    }
+
+    if (reply->type == REDIS_REPLY_INTEGER ) {
```
Comment (regit) [inline src/util-logopenfile.c]: Please put test in debug preprocessor test to avoid something useless. In non debug mode SCLogDebug is just replace by an empty line.

File: src/util-logopenfile.c [hunk 10]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+        SCLogDebug("redis reply: %lld\n", reply->integer);
+    }
+
+    if (reply->type == REDIS_REPLY_STRING && reply->str) {
```
Comment (regit) [inline src/util-logopenfile.c]: idem

File: src/util-logopenfile.c [hunk 11]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+		return -1;
+	}
+
+	SCLogRedisContext * ctx = log_ctx->redis;
+	const char *redis_server = log_ctx->redis_setup.server;
+	int redis_port = log_ctx->redis_setup.port;
+
+#if HAVE_LIBEVENT == 1
+    if (log_ctx->redis_setup.async) {
+
```
Comment (regit) [inline src/util-logopenfile.c]: Too much use of empty line, it makes code hard to read.

File: src/util-logopenfile.c [hunk 12]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+            redisAsyncSetDisconnectCallback(c,RedisDisconnectCallback);
+            SCLogInfo("Connection to redis server [%s]:[%d] will use async.",
+                    redis_server, redis_port);
+
+            redisAsyncHandleWrite(c);
+
+            ctx->async = c;
+        }
+    } else
+#endif
```
Comment (regit) [inline src/util-logopenfile.c]: This construct with `#if` make indent looks weird.

File: src/util-logopenfile.c [hunk 13]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+	{
+        /* SYNCHRONOUS */
+        if (ctx->sync != NULL)  {
+            redisFree(ctx->sync);
+        }
+
+        ctx->sync = NULL;
```
Comment (regit) [inline src/util-logopenfile.c]: This can be moved in preceeding if

File: src/util-logopenfile.c [hunk 14]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+
+        redisContext *c = redisConnect(redis_server, redis_port);
+        if ( c != NULL && c->err) {
```
Comment (regit) [inline src/util-logopenfile.c]: '( c' no space please

File: src/util-logopenfile.c [hunk 15]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
+            SCLogError(SC_ERR_SOCKET, "Error connecting to redis server: [%s] !", c->errstr);
+            redisFree(c);
+            ctx->sync = NULL;
```
Comment (regit) [inline src/util-logopenfile.c]: Useless here.

Comment (regit) [inline src/util-logopenfile.c]: why not directly use ctx->sync and remove that c ?

File: src/util-logopenfile.c [hunk 16]
```diff
@@ -336,29 +361,253 @@ int SCConfLogReopen(LogFileCtx *log_ctx)
...
     }
+
+    log_ctx->redis = ctx;
     log_ctx->redis_setup.tried = 0;
     log_ctx->redis_setup.batch_count = 0;
+    return 0;
 }
 
+
+/** \brief configure and initializes redis output logginh
```
Comment (regit) [inline src/util-logopenfile.c]: logging

File: src/util-logopenfile.c [hunk 17]
```diff
@@ -410,59 +669,28 @@ int SCConfLogOpenRedis(ConfNode *redis_node, LogFileCtx *log_ctx)
     } else {
         log_ctx->redis_setup.command = redis_publish_cmd;
         if (!log_ctx->redis_setup.command) {
-            SCLogError(SC_ERR_MEM_ALLOC, "Unable to allocate redis key command");
+            SCLogError(SC_ERR_MEM_ALLOC, "Unable to allocate redis key command!");
```
Comment (regit) [inline src/util-logopenfile.c]: Please next time, no cosmetic like that on such a huge patch

File: src/util-logopenfile.c [hunk 18]
```diff
@@ -538,19 +766,62 @@ int LogFileFreeCtx(LogFileCtx *lf_ctx)
...
+ */
+static int LogFileWriteRedis(LogFileCtx *file_ctx, const char *string, size_t string_len)
 {
     if (file_ctx->redis == NULL) {
         SCConfLogReopenRedis(file_ctx);
         if (file_ctx->redis == NULL) {
             return -1;
         } else {
-            SCLogInfo("Reconnected to redis server");
+            SCLogInfo("Reconnected to redis server.");
```
Comment (regit) [inline src/util-logopenfile.c]: Please..

File: src/util-logopenfile.c [hunk 19]
```diff
@@ -538,19 +766,62 @@ int LogFileFreeCtx(LogFileCtx *lf_ctx)
...
         }
     }
-    /* TODO go async here ? */
+
+	SCLogRedisContext * ctx = file_ctx->redis;
```
Comment (regit) [inline src/util-logopenfile.c]: Indent seems wrong.

File: src/util-logopenfile.c [hunk 20]
```diff
@@ -538,19 +766,62 @@ int LogFileFreeCtx(LogFileCtx *lf_ctx)
...
+            redisAsyncCommand(redis_async,
+                    SCRedisAsyncCommandCallback,
+                    file_ctx,
+                    "%s %s %s",
+                    file_ctx->redis_setup.command,
+                    file_ctx->redis_setup.key,
+                    string);
+
+            redisAsyncHandleWrite(redis_async);
+            redisAsyncHandleRead(redis_async);
```
Comment (regit) [inline src/util-logopenfile.c]: So we will check for write answer after write. Should not it be also done when closing the connection to redis ?

File: src/util-logopenfile.c [hunk 21]
```diff
@@ -538,19 +766,62 @@ int LogFileFreeCtx(LogFileCtx *lf_ctx)
...
+
+        }
+        return 0;
+    }
+#endif
```
Comment (regit) [inline src/util-logopenfile.c]: This construct is not nice. It would be better to have an async and a sync function defined inline and then call them in a nice if else construct.

File: suricata.yaml.in [hunk 22]
```diff
@@ -144,6 +144,7 @@ outputs:
       #redis:
       #  server: 127.0.0.1
       #  port: 6379
+      #  async: true
```
Comment (regit) [inline suricata.yaml.in]: explanation here may help users.

File: src/util-logopenfile.c [hunk 23]
```diff
@@ -592,26 +863,35 @@ static int  LogFileWriteRedis(LogFileCtx *file_ctx, const char *string, size_t s
...
-                SCConfLogReopenRedis(file_ctx);
-                break;
-            case REDIS_REPLY_INTEGER:
-                SCLogDebug("Redis integer %lld", reply->integer);
-                break;
-            default:
-                SCLogError(SC_ERR_INVALID_VALUE,
-                        "Redis default triggered with %d", reply->type);
+        /* We may lose the reply if disconnection happens*/
+        if (reply)  {
```
Comment (regit) [inline src/util-logopenfile.c]: This is handling sync code and fixes a real problem. Well done but it should have been a separate patch so we can backport it to stable version.

File: doc/userguide/configuration/suricata-yaml.rst [hunk 24]
```diff
@@ -267,6 +267,7 @@ integration with 3rd party tools like logstash.
       #redis:
       #  server: 127.0.0.1
       #  port: 6379
+      #  async: true ## if redis replies are read asynchronously
```
Comment (regit) [inline doc/userguide/configuration/suricata-yaml.rst]: This fixes comment made in previous commit.

**Comment by inliniac** [comment]
I think it would be cleaner to move the redis code into it's own file as well. Something like util-log-redis.[ch] or something.

**Comment by fooinha** [comment]
Yes, I'll move this code to new files. I also did a output for kafka, and the merge would be messy.
@regit , thanx for your review. I'll try to put your recommendations in a new PR version.
